iceberg/streams/iceberg/file_reader_builder.cpp

```diff
@@ -0,0 +1,126 @@
+#include "iceberg/streams/iceberg/file_reader_builder.h"
+
+namespace iceberg {
+
+namespace {
+
+void InsertOrFail(std::unordered_set<int>& set, int x) {
+  Ensure(!set.contains(x), "Field ids are not unique");
+  set.insert(x);
+}
+
+std::unordered_map<int, int> MakeIndexToFieldIdMap(const parquet::schema::GroupNode* node,
+                                                   const std::optional<std::string>& schema_name_mapping) {
+  if (!schema_name_mapping.has_value()) {
+    return {};
+  }
+  std::unordered_set<int> field_ids_set;
+  std::optional<SchemaNameMapper> schema_name_mapper;
+  std::optional<SchemaNameMapper::Node> root_node;
+  std::unordered_map<int, int> index_to_field_id;
+  for (int i = 0; i < node->field_count(); ++i) {
+    if (node->field(i)->field_id() >= 0) {
+      InsertOrFail(field_ids_set, node->field(i)->field_id());
+      continue;
+    }
+    if (!schema_name_mapper.has_value()) {
+      schema_name_mapper = SchemaNameMapper(*schema_name_mapping);
```

Comment:
It's not very good to call `SchemaNameMapper` over every node. Could you make 
`std::unordered_map<int, int> MakeIndexToFieldIdMap(const parquet::schema::GroupNode* node, <type> schema_name_mapper)` and call `SchemaNameMapper` once?